File: mini-variational-principle/src/second_variation.c
/**
 * second_variation.c -- Second Variation Theory (L4-L6)
 * Legendre, Jacobi, Weierstrass conditions, Hilbert integral,
 * sufficient conditions, Ritz/Galerkin methods, Morse theory.
 */
#include "second_variation.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <float.h>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

bool legendre_condition(double v) { return v >= -1e-12; }
bool strong_legendre(double v) { return v > 1e-12; }

void verify_legendre_along_extremal(
    LagrangianDensity d2Fyp2, const double *x, const double *y,
    const double *yp, int n, bool *ok, double *mn, double *mx, double *pv)
{
    if (!d2Fyp2||!x||!y||!yp||n<2) { if(ok)*ok=false; if(mn)*mn=0; if(mx)*mx=0; return; }
    bool p=true; double a=DBL_MAX,b=-DBL_MAX;
    for (int i=0;i<n;i++) {
        double P=d2Fyp2(x[i],y[i],yp[i]); if(pv) pv[i]=P;
        if(P<a)a=P; if(P>b)b=P; if(P<-1e-10) p=false;
    }
    if(ok)*ok=p; if(mn)*mn=a; if(mx)*mx=b;
}

int find_conjugate_points(double a, double b,
    LagrangianDensity P_func, LagrangianDensity Q_func,
    const double *xe, const double *ye, const double *ype,
    int ng, double *cout, int mxp)
{
    if (!P_func||!Q_func||!xe||!ye||!ype||ng<3) return 0;
    double dx=(b-a)/(double)(ng-1);
    double *u=malloc((size_t)ng*sizeof(double));
    double *up=malloc((size_t)ng*sizeof(double));
    if (!u||!up) { free(u); free(up); return 0; }
    u[0]=0.0; up[0]=1.0;
    for (int i=0;i<ng-1;i++) {
        double xm=xe[i]+0.5*dx, P=P_func(xm,0.0,0.0), Q=Q_func(xm,0.0,0.0);
        if (fabs(P)<1e-15) P=1e-15;
        up[i+1]=up[i]+dx*Q*u[i]/P; u[i+1]=u[i]+dx*up[i];
    }
    int np=0;
    for (int i=1;i<ng&&np<mxp;i++) {
        if (u[i-1]*u[i]<0.0&&cout) {
            double t=fabs(u[i-1])/(fabs(u[i-1])+fabs(u[i]));
            cout[np++]=xe[i-1]+t*dx;
        }
    }
    free(u); free(up); return np;
}

bool jacobi_condition(const double *cp, int n, double a, double b) {
    for (int i=0;i<n;i++) if (cp[i]>a&&cp[i]<b) return false;
    return true;
}
double first_conjugate_point(const double *cp, int n, double a) {
    double f=INFINITY;
    for (int i=0;i<n;i++) if (cp[i]>a&&cp[i]<f) f=cp[i];
    return f;
}
void solve_jacobi_equation(double a, double b,
    LagrangianDensity Pf, LagrangianDensity Qf,
    const double *xe, const double *ye, const double *ype,
    int ng, double u0p, double *uo, double *upo)
{
    if (!Pf||!Qf||!xe||!uo||!upo||ng<2) return;
    double dx=(b-a)/(double)(ng-1); uo[0]=0.0; upo[0]=u0p;
    for (int i=0;i<ng-1;i++) {
        double xm=xe[i]+0.5*dx, P=Pf(xm,0.0,0.0), Q=Qf(xm,0.0,0.0);
        if (fabs(P)<1e-15) P=1e-15;
        upo[i+1]=upo[i]+dx*Q*uo[i]/P; uo[i+1]=uo[i]+dx*upo[i];
    }
}

double weierstrass_excess(LagrangianDensity F, LagrangianDensity dFdp,
    double x, double y, double po, double p) {
    if (!F||!dFdp) return 0.0;
    return F(x,y,p)-F(x,y,po)-(p-po)*dFdp(x,y,po);
}
bool weierstrass_condition(LagrangianDensity F, LagrangianDensity dFdp,
    double x, double y, double po, const double *pr, int np) {
    if (!F||!dFdp||!pr||np<1) return false;
    for (int i=0;i<np;i++)
        if (weierstrass_excess(F,dFdp,x,y,po,pr[i])<-1e-10) return false;
    return true;
}
bool verify_weierstrass_along_extremal(LagrangianDensity F,
    LagrangianDensity dFdp, const double *x, const double *y,
    const double *yp, int nn, const double *pr, int np) {
    if (!F||!dFdp||!x||!y||!yp||nn<1) return false;
    for (int i=0;i<nn;i++)
        if (!weierstrass_condition(F,dFdp,x[i],y[i],yp[i],pr,np)) return false;
    return true;
}

double hilbert_invariant_integral(LagrangianDensity F, LagrangianDensity dFdp,
    const double *x, const double *y, const double *yp,
    double (*pf)(double,double), int n) {
    if (!F||!dFdp||!x||!y||!yp||!pf||n<2) return 0.0;
    double I=0.0,dx=x[1]-x[0];
    for (int i=0;i<n;i++) {
        double p=pf(x[i],y[i]), w=(i==0||i==n-1)?0.5:1.0;
        I+=w*(F(x[i],y[i],p)+(yp[i]-p)*dFdp(x[i],y[i],p))*dx;
    }
    return I;
}
double hilbert_integral_test(LagrangianDensity F, LagrangianDensity dFdp,
    double (*pf)(double,double), const double *x,
    const double *y1, const double *yp1,
    const double *y2, const double *yp2, int n) {
    double I1=hilbert_invariant_integral(F,dFdp,x,y1,yp1,pf,n);
    double I2=hilbert_invariant_integral(F,dFdp,x,y2,yp2,pf,n);
    return fabs(I1-I2);
}

SufficientConditions check_sufficient_conditions(
    LagrangianDensity F, LagrangianDensity dFdp, LagrangianDensity d2Fyp2,
    LagrangianDensity Pf, LagrangianDensity Qf,
    const double *x, const double *y, const double *yp,
    int n, const double *pr, int np)
{
    SufficientConditions sc; memset(&sc,0,sizeof(sc));
    bool lp=false; double mn,mx;
    double *pv=malloc((size_t)n*sizeof(double));
    if (pv) {
        verify_legendre_along_extremal(d2Fyp2,x,y,yp,n,&lp,&mn,&mx,pv);
        sc.legendre_ok=lp&&(mn>1e-12); free(pv);
    }
    double cp[100];
    int nc=find_conjugate_points(x[0],x[n-1],Pf,Qf,x,y,yp,n,cp,100);
    sc.jacobi_ok=jacobi_condition(cp,nc,x[0],x[n-1]);
    sc.weierstrass_ok=verify_weierstrass_along_extremal(F,dFdp,x,y,yp,n,pr,np);
    sc.sufficient=sc.legendre_ok&&sc.jacobi_ok&&sc.weierstrass_ok;
    return sc;
}

int ritz_method(LagrangianDensity d2Fyp2, LagrangianDensity d2Fy2,
    double a, double b, int nb, double (**ba)(double),
    double (**bd)(double), int nq, double *co)
{
    if (!d2Fyp2||!d2Fy2||!ba||!bd||!co||nb<1||nq<2) return -1;
    double *A=calloc((size_t)(nb*nb),sizeof(double));
    double *fv=calloc((size_t)nb,sizeof(double));
    if (!A||!fv) { free(A); free(fv); return -1; }
    double dx=(b-a)/(double)(nq-1);
    for (int k=0;k<nq;k++) {
        double x=a+(double)k*dx, w=(k==0||k==nq-1)?0.5:1.0;
        double P=d2Fyp2(x,0.0,0.0), Q=d2Fy2(x,0.0,0.0);
        for (int i=0;i<nb;i++)
            for (int j=0;j<nb;j++)
                A[i*nb+j]+=w*(P*bd[i](x)*bd[j](x)+Q*ba[i](x)*ba[j](x))*dx;
    }
    for (int iter=0;iter<1000;iter++) {
        double md=0.0;
        for (int i=0;i<nb;i++) {
            double s=0.0;
            for (int j=0;j<nb;j++) if (j!=i) s+=A[i*nb+j]*co[j];
            double nc=-s/A[i*nb+i];
            double d=fabs(nc-co[i]); if (d>md) md=d; co[i]=nc;
        }
        if (md<1e-12) break;
    }
    free(A); free(fv); return 0;
}

int galerkin_method(double (*el_op)(double(*u)(double),double x),
    double a, double b, int nb, double (**ba)(double),
    int nq, double *co)
{
    if (!el_op||!ba||!co||nb<1||nq<2) return -1;
    double *A=calloc((size_t)(nb*nb),sizeof(double));
    if (!A) return -1;
    double dx=(b-a)/(double)(nq-1);
    for (int k=0;k<nq;k++) {
        double x=a+(double)k*dx, w=(k==0||k==nq-1)?0.5:1.0;
        for (int i=0;i<nb;i++) {
            double ev=el_op(ba[i],x);
            for (int j=0;j<nb;j++) A[i*nb+j]+=w*ev*ba[j](x)*dx;
        }
    }
    for (int iter=0;iter<500;iter++) {
        double md=0.0;
        for (int i=0;i<nb;i++) {
            double s=0.0;
            for (int j=0;j<nb;j++) if (j!=i) s+=A[i*nb+j]*co[j];
            double nc=-s/A[i*nb+i]; double d=fabs(nc-co[i]);
            if (d>md) md=d; co[i]=nc;
        }
        if (md<1e-10) break;
    }
    free(A); return 0;
}

int morse_index(const double *cp, int n, double a, double b) {
    int c=0; for (int i=0;i<n;i++) if (cp[i]>a&&cp[i]<b) c++; return c;
}
int count_jacobi_zeros(const double *u, int n) {
    if (!u||n<2) return 0;
    int c=0; for (int i=1;i<n;i++) if (u[i-1]*u[i]<0.0) c++; return c;
}

File: mini-variational-principle/src/second_variation.jl
# second_variation.jl — 二阶变分理论: Legendre条件, Jacobi条件, Weierstrass, Hilbert
# 参考: Gelfand & Fomin Ch.5, Bolza "Lectures on the Calculus of Variations"

using LinearAlgebra

# ============================================================
# Legendre 条件 (二阶必要条件)
# ============================================================
"Legendre 条件: ∂²F/∂y''² ≥ 0 (弱必要条件)"
legendre_condition(d2Fdyp2::Float64) = d2Fdyp2 >= -1e-12

"增强 Legendre 条件: ∂²F/∂y''² > 0 (强必要条件)"
strong_legendre(d2Fdyp2::Float64) = d2Fdyp2 > 1e-12

"Legendre 条件沿极值曲线验证"
function verify_legendre_along_extremal(d2Fdyp2::Function, y_sol::Function, yp_sol::Function, a::Float64, b::Float64; n=200)
    xs=range(a,b,length=n); vals=[d2Fdyp2(x,y_sol(x),yp_sol(x)) for x in xs]
    all(v->v>=-1e-10, vals), minimum(vals), xs, vals
end

# ============================================================
# Jacobi 条件 (共轭点)
# ============================================================
"找共轭点: 解 Jacobi 方程 (P u''''+ Q u = 0), 其中 P=∂²F/∂y''², Q=∂²F/∂y²-d/dx(∂²F/∂y∂y'')"
function find_conjugate_points(a::Float64, b::Float64, P::Function, Q::Function; n=1000)
    dx=(b-a)/(n-1); u=zeros(n); du=zeros(n)
    u[1]=0.0; du[1]=1.0
    for i in 1:n-1; xm=a+(i-0.5)*dx
        du[i+1]=du[i]+dx*(-Q(xm)*u[i])
        u[i+1]=u[i]+dx*(du[i]/max(P(xm),1e-15))
    end
    pts=Float64[]
    for i in 2:n; u[i-1]*u[i]<0 && push!(pts,a+(i-1)*dx-dx*abs(u[i-1])/(abs(u[i-1])+abs(u[i]))); end
    pts
end

"Jacobi 条件: (a,b) 内无共轭点"
function jacobi_condition(pts::Vector{Float64}, a::Float64, b::Float64)
    isempty(filter(x->a<x<b, pts))
end

"共轭点最近的 x 坐标"
function first_conjugate_point(pts::Vector{Float64}, a::Float64)
    filtered=filter(x->x>a, pts)
    isempty(filtered) ? Inf : minimum(filtered)
end

# ============================================================
# Weierstrass 条件
# ============================================================
"Weierstrass 过剩函数: E(x,y,y''_opt, y''_test) = F(x,y,y''_test)-F(x,y,y''_opt)-(y''_test-y''_opt)∂F/∂y''|_{opt}"
function weierstrass_excess(F::Function, dFdyp::Function, x::Float64, y::Float64, yp_opt::Float64, yp_test::Float64)
    F(x,y,yp_test)-F(x,y,yp_opt)-(yp_test-yp_opt)*dFdyp(x,y,yp_opt)
end

"Weierstrass 必要条件: E(x,y,p_opt, p) ≥ 0 ∀p"
function weierstrass_condition(F::Function, dFdyp::Function, x::Float64, y::Float64, yp_opt::Float64, p_range::AbstractRange)
    all(weierstrass_excess(F,dFdyp,x,y,yp_opt,p) >= -1e-10 for p in p_range)
end

"沿极值曲线验证 Weierstrass 条件"
function verify_weierstrass_along_extremal(F::Function, dFdyp::Function, y_sol::Function, yp_sol::Function, a::Float64, b::Float64, p_range::AbstractRange; n=100)
    xs=range(a,b,length=n)
    all(weierstrass_condition(F,dFdyp,x,y_sol(x),yp_sol(x),p_range) for x in xs)
end

# ============================================================
# Hilbert 不变积分
# ============================================================
"Hilbert 不变积分: ∫[F(x,y,p)+(y''-p)∂F/∂p]dx (与路径无关)"
function hilbert_invariant_integral(F::Function, dFdyp::Function, y::Function, yp::Function, p_field::Function, a::Float64, b::Float64; n=500)
    dx=(b-a)/n; I=0.0
    for i in 1:n; x=a+(i-0.5)*dx
        p=p_field(x,y(x))
        I+=(F(x,y(x),p)+(yp(x)-p)*dFdyp(x,y(x),p))*dx
    end; I
end

"构造斜率场 p(x,y) 满足可积条件 (Hamilton-Jacobi)"
function slope_field_from_extremal(y_sol::Function, yp_sol::Function, a::Float64, b::Float64; n=100)
    xs=range(a,b,length=n)
    x->begin
        idx=argmin(abs.(xs.-x))
        yp_sol(xs[max(1,min(n,idx))])
    end
end

# ============================================================
# 充分条件
# ============================================================
"强极小值的充分条件: Legendre + Jacobi + Weierstrass"
function sufficient_conditions_strong(d2Fdyp2::Function, y_sol::Function, yp_sol::Function, P::Function, Q::Function, F::Function, dFdyp::Function, a::Float64, b::Float64; p_range=-10:0.1:10, n=200)
    leg_ok,_,_,_=verify_legendre_along_extremal(d2Fdyp2,y_sol,yp_sol,a,b;n=n)
    cj_pts=find_conjugate_points(a,b,P,Q;n=500)
    jac_ok=jacobi_condition(cj_pts,a,b)
    wei_ok=verify_weierstrass_along_extremal(F,dFdyp,y_sol,yp_sol,a,b,p_range;n=n)
    (Legendre=leg_ok, Jacobi=jac_ok, Weierstrass=wei_ok, Sufficient=leg_ok&&jac_ok&&wei_ok)
end

# ============================================================
# Ritz 方法 (直接法)
# ============================================================
"Ritz 法: 用有限基函数近似求解变分问题"
function ritz_method(F::Function, basis::Vector{Function}, basis_deriv::Vector{Function}, a::Float64, b::Float64; n=500, lambda_reg=0.0)
    m=length(basis); A=zeros(m,m); b_vec=zeros(m); dx=(b-a)/n
    for k in 1:n; x=a+(k-0.5)*dx
        for i in 1:m, j in 1:m; A[i,j]+=basis_deriv[i](x)*basis_deriv[j](x)*dx; end
        for i in 1:m; b_vec[i]+=basis[i](x)*dx; end
    end
    A_reg=A+lambda_reg*Matrix{Float64}(I,m,m)
    c=A_reg\b_vec; (x->sum(c[i]*basis[i](x) for i in 1:m), c)
end

"Galerkin 法: 加权残差 (用于 PDE → ODE)"
function galerkin_method(el_operator::Function, basis::Vector{Function}, a::Float64, b::Float64; n=500)
    m=length(basis); A=zeros(m,m); b_vec=zeros(m); dx=(b-a)/n
    for k in 1:n; x=a+(k-0.5)*dx
        for i in 1:m, j in 1:m; A[i,j]+=el_operator(basis[j],x)*basis[i](x)*dx; end
        for i in 1:m; b_vec[i]+=0.0*dx; end  # 齐次问题
    end
    c=A\b_vec; (x->sum(c[i]*basis[i](x) for i in 1:m), c)
end

"Kantorovich 法 (半离散化: 仅部分积分)"
function kantorovich_method(F::Function, x_basis::Vector{Function}, y_range::Tuple{Float64,Float64}; nx=100, ny=50)
    # 2D → 1D 的降维方法
    m=length(x_basis); c_funcs=[y->0.0 for _ in 1:m]  # placeholder
    c_funcs
end

# ============================================================
# Morse 理论 (极小极大的计数)
# ============================================================
"共轭点的 Morse 指数: (a,b) 内共轭点的个数"
function morse_index(conjugate_points::Vector{Float64}, a::Float64, b::Float64)
    count(pt->a<pt<b, conjugate_points)
end

"共轭点间的零点计数"
function count_jacobi_zeros(u_sol::Vector{Float64})
    count=0
    for i in 2:length(u_sol); u_sol[i-1]*u_sol[i]<0 && (count+=1); end
    count
end

export legendre_condition, strong_legendre, verify_legendre_along_extremal
export find_conjugate_points, jacobi_condition, first_conjugate_point
export weierstrass_excess, weierstrass_condition, verify_weierstrass_along_extremal
export hilbert_invariant_integral, slope_field_from_extremal, sufficient_conditions_strong
export ritz_method, galerkin_method, kantorovich_method
export morse_index, count_jacobi_zeros
